### PR TITLE
preload sender and check for errors

### PR DESCRIPTION
**Pull request by hexoscott**

### CODE REVIEW

**Review by revitteth — APPROVED** [review]

File: zk/stages/stage_sequence_execute_transactions.go [hunk 1]
```diff
@@ -96,12 +99,23 @@ func extractTransactionsFromSlot(slot *types2.TxsRlp) ([]types.Transaction, []co
...
+				"error", err,
+				"id", slot.TxIds[idx])
 			toRemove = append(toRemove, slot.TxIds[idx])
 			continue
 		}
-		var sender common.Address
-		copy(sender[:], slot.Senders.At(idx))
+
+		// now attempt to recover the sender
+		sender, err := signer.SenderWithContext(cryptoContext, transaction)
```
Comment (doutv) [inline zk/stages/stage_sequence_execute_transactions.go:110]: pprof shows that this line takes 27% of total time. I run sequencer and benchmark

Comment (doutv) [reply zk/stages/stage_sequence_execute_transactions.go:110]: `sender, err := signer.SenderWithContext(cryptoContext, transaction)`

Comment (doutv) [reply zk/stages/stage_sequence_execute_transactions.go:110]: Related issue: 
https://github.com/okx/xlayer-erigon/issues/211

Comment (doutv) [reply zk/stages/stage_sequence_execute_transactions.go:110]: Why attempt to recover sender from public key? there should be sender address in the transaction

File: zk/stages/stage_sequence_execute_transactions.go [hunk 2]
```diff
@@ -84,10 +85,12 @@ func getLimboTransaction(ctx context.Context, cfg SequenceBlockCfg, txHash *comm
 	return transactions, nil
 }
 
-func extractTransactionsFromSlot(slot *types2.TxsRlp) ([]types.Transaction, []common.Hash, []common.Hash, error) {
+func extractTransactionsFromSlot(slot *types2.TxsRlp, currentHeight uint64, cfg SequenceBlockCfg) ([]types.Transaction, []common.Hash, []common.Hash, error) {
 	ids := make([]common.Hash, 0, len(slot.TxIds))
 	transactions := make([]types.Transaction, 0, len(slot.Txs))
 	toRemove := make([]common.Hash, 0)
+	signer := types.MakeSigner(cfg.chainConfig, currentHeight, 0)
+	cryptoContext := secp256k1.ContextForThread(1)
```
Comment (doutv) [inline zk/stages/stage_sequence_execute_transactions.go:93]: Use only 1 context for all transactions. This is sequential computation. Why not parallel?

**Comment by doutv** [comment]
@hexoscott @revitteth 
I need your help! Please check my review comment. 

I want to know more background about this PR. 
- Why recover sender address from public key? Is it necessary? 

Some optimization ideas:
- Read sender address in slot first, when read failed, fallback to recover sender address
- Parallel cryptoContext, recover sender address in parallel

My draft PR just remove "recover sender address from public key" and read sender address from transaction slot. Sequencer can be **50% faster** under our benchmark.

File: zk/stages/stage_sequence_execute_transactions.go (hunk 1 above)
Comment (hexoscott) [reply zk/stages/stage_sequence_execute_transactions.go:110]: The sender always needs recovering at some point, it isn't part of the RLP for a transaction.  Once decoded it is cached on the transaction but until then something needs to recover it.

**Comment by hexoscott** [comment]
The context for this change was that some chains were receiving spurious transactions that looked fine but the R,S,V values on the transaction would not allow for the sender to be recovered which caused an error.  So, we pre-load the sender now and check for this error ahead of time to ensure this doesn't happen.  We need to do it at this point in the process otherwise the transaction has already been added to the block and it's too late to do anything about it.

Fixing the issue was the priority at the time, if there is degraded performance then I'll raise a new issue with your points above.

**Pull request merged**